src/hooks/useCrop.ts

```diff
@@ -0,0 +1,377 @@
+      updatedLeft = oldCropArea.left + diffX;
+      updatedWidth = oldCropArea.width - diffX;
+      break;
+
+    case 'move':
+      updatedTop = oldCropArea.top + diffY;
+      updatedLeft = oldCropArea.left + diffX;
+      break;
+
+    default:
+      break;
+  }
+
+  const left = clamp(updatedLeft, imgArea.left, (imgArea.width - lastPoint.width + imgArea.left));
+  const top = clamp(updatedTop, imgArea.top, (imgArea.height - lastPoint.height + imgArea.top));
+  const width = clamp(updatedWidth, 0, imgArea.width);
+  const height = clamp(updatedHeight, 0, imgArea.height);
+
+  return {
+    left,
+    top,
+    width,
+    height,
+  };
+}
+
+export interface IDrawArea {
+  left: number
+  top: number
+  width: number
+  height: number
+}
+
+interface ILastPoint {
+  posX: number,
+  posY: number,
+  width: number
+  height: number
+}
+
+export type ICropState = {
+  cursorStyle : ICursorStyles,
+  resizeCursorStyle: ICursorStyles,
+  isResizing: boolean,
+  cropDraw: IDrawArea,
+  canvasDraw: IDrawArea,
+  imageDraw: IDrawArea,
+  lastCursorPoint: ILastPoint,
+}
+
+const UPDATE_CURSOR = 'UPDATE_CURSOR';
+const RESIZE_CROP_AREA = 'RESIZE_CROP_AREA';
+const START_RESIZE = 'START_RESIZE';
+const END_RESIZE = 'END_RESIZE';
+const SET_DRAW_AREAS = 'SET_DRAW_AREAS';
+
+type ICropReducerActions = (
+  IUpdateCursor
+  | IStartResize
+  | IResizeCropArea
+  | IEndResize
+  | SET_DRAW_AREAS
+);
+
+interface IUpdateCursor {
+  type: 'UPDATE_CURSOR'
+  data: {
+    cursorStyle: ICursorStyles
+  } 
+}
+
+interface IStartResize {
+  type: 'START_RESIZE'
+  data: {
+    resizeCursor: ICursorStyles,
+    lastPosX: number,
+    lastPosY: number,
+    lastHeight: number,
+    lastWidth: number,
+  }
+}
+
+interface IResizeCropArea {
+  type: 'RESIZE_CROP_AREA'
+  data: {
+    newPosX: number
+    newPosY: number
+    newHeight: number
+    newWidth: number
+  }
+}
+
+interface IEndResize {
+  type: 'END_RESIZE'
+}
+
+interface SET_DRAW_AREAS {
+  type: 'SET_DRAW_AREAS',
+  data: {
+    canvasDraw: IDrawArea,
+    imageDraw: IDrawArea,
+    cropDraw: IDrawArea,
+  }
+}
+
+const defaultCropValues : ICropState = {
+  cursorStyle: 'default',
+  resizeCursorStyle: 'default',
+  isResizing: false,
+  cropDraw: {
+    left: 0,
+    top: 0,
+    height: 100,
+    width: 100,
+  },
+  canvasDraw: {
+    left: 0,
+    top: 0,
+    height: 100,
+    width: 100,
+  },
+  imageDraw: {
+    left: 0,
+    top: 0,
+    height: 100,
+    width: 100,
+  },
+  lastCursorPoint: {
+    posX: 0,
+    posY: 0,
+    height: 100,
+    width: 100,
+  }
+};
+
+
+function cropReducer(state: ICropState, action: ICropReducerActions){
+  switch(action.type) {
+
+    case SET_DRAW_AREAS :
+      return {...state,
+        canvasDraw: action.data.canvasDraw,
+        imageDraw: action.data.imageDraw,
+        cropDraw: action.data.cropDraw,
+      };
+
+    case UPDATE_CURSOR:
+      return {...state, cursorStyle: action.data.cursorStyle};
+
+    case START_RESIZE:
+      {
```

Comment:
This is in other to be able to declare a variable inside the case, if not it will complain.

https://medium.com/@e_himmelfarb/use-curly-braces-with-es6-let-and-const-in-switch-blocks-react-redux-reducers-c0b01b37d748